lib/ukdebug/arch/arm64/gdbsup.c

```diff
@@ -0,0 +1,132 @@
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "gdbsup.h"
+#include "../../gdbstub.h"
+
+#include <uk/arch/traps.h>
+#include <uk/arch/lcpu.h>
+#include <uk/essentials.h>
+#include <uk/nofault.h>
+#include <uk/bitops.h>
+#include <uk/compiler.h>
+
+#define MDSCR_EL1_SS		UK_BIT(0)
+#define MDSCR_EL1_KDE		UK_BIT(13)
+
+#define SPSR_EL1_SS		UK_BIT(21)
+#define SPSR_EL1_D		UK_BIT(9)
+
+#define ESR_EL1_EC_BRK64	0x3c
+#define BRK_OPCODE		0xd4200000UL
+
+static void gdb_arch_enable_single_step(struct __regs *regs)
+{
+	__sz mdscr = SYSREG_READ(mdscr_el1);
+
+	mdscr |= MDSCR_EL1_SS;
+	mdscr |= MDSCR_EL1_KDE;
+
+	SYSREG_WRITE(mdscr_el1, mdscr);
+
+	regs->spsr_el1 |= SPSR_EL1_SS;
+}
+
+static void gdb_arch_disable_single_step(struct __regs *regs)
+{
+	__sz mdscr = SYSREG_READ(mdscr_el1);
+
+	mdscr &= ~MDSCR_EL1_SS;
+	mdscr &= ~MDSCR_EL1_KDE;
+
+	SYSREG_WRITE(mdscr_el1, mdscr);
+
+	regs->spsr_el1 &= ~SPSR_EL1_SS;
+}
+
+/* We get here via traps raised by the platform
+ * TODO: Once the crash screen PR is merged, crashes can land us here
+ * too, if the "enter debugger on crash" feature is enabled.
+ */
+static int gdb_arch_dbg_trap(int errnr, struct __regs *regs)
+{
+	int r;
+
+	gdb_arch_disable_single_step(regs);
+
+	r = gdb_dbg_trap(errnr, regs);
+	if (r < 0) {
+		return r;
+	} else if (r == GDB_DBG_STEP) { /* Single step */
+		gdb_arch_enable_single_step(regs);
+
+		regs->spsr_el1 &= ~SPSR_EL1_D;
+	}
+
+	return UK_EVENT_HANDLED;
+}
+
+static int gdb_arch_is_brk(unsigned long pc)
+{
+	__u32 opcode;
+
+	if (!uk_nofault_memcpy((void *)&opcode, (void *)pc, sizeof(opcode),
+			       UK_NOFAULTF_NOPAGING))
+		return 0;
+
+	return ((opcode & BRK_OPCODE) == BRK_OPCODE);
```

Comment:
I didn't know we had a bool type

Reply:
As `libukdebug` depends on `HAVE_LIBC` I guess it's okay to include `stdbool.h`. See also the `Return Values`  section in https://unikraft.org/docs/contributing/coding-conventions, although this may need amending given that, as you said, there is no kernel `__bool` type.